Divrem speedup and tests

r7rohan opened the pull request:
[#22219](https://github.com/JuliaLang/julia/issues/22219#issue-233492600)
- Replaced rem call with `a-d*b`. Reduced div calls from 2-->1 (code in: `base\div.jl`)
Int-Int : Works for all rounding modes and speeds up by a factor of 1.6x
Floats - Abstract Floats, Abstract Irrational : Works but there is marginal error in Float64, BigFloat. (Not added)
Rationals - `a-d*b` is more expensive than normal div and rem calls. (Not added)

- Also added tests for divrem in rounding modes in `test\numbers.jl`

Passes all the tests for div, divrem, rem ` SUCCESS`
Also speeds divrem by a factor of 2
```
function divrem!(a, b, c, d)
    for i = 1:100_000
        c[i], d[i] = divrem(a[i], b[i])
    end
end

function divrem2!(a, b, c, d)
    for i = 1:100_000
        c[i] = div(a[i], b[i])
        d[i] = rem(a[i], b[i])
    end
end

function divrem3!(a, b, c, d)
    for i = 1:100_000
        c[i] = div(a[i], b[i])
        d[i] = a[i] - b[i] * c[i]
    end
end

a, b = rand(0:500_000, 100_000), rand(1:250_000, 100_000);
c, d = zeros(Int, 100_000), zeros(Int, 100_000)


julia> @benchmark divrem!($a,$b,$c,$d) samples=100
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     834.000 μs (0.00% GC)
  median time:      837.300 μs (0.00% GC)
  mean time:        851.657 μs (0.00% GC)
  maximum time:     1.048 ms (0.00% GC)
  --------------
  samples:          100
  evals/sample:     1

julia> @benchmark divrem2!($a,$b,$c,$d) samples=100
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     1.479 ms (0.00% GC)
  median time:      1.480 ms (0.00% GC)
  mean time:        1.499 ms (0.00% GC)
  maximum time:     1.716 ms (0.00% GC)
  --------------
  samples:          100
  evals/sample:     1

julia> @benchmark divrem3!($a,$b,$c,$d) samples=100
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     839.100 μs (0.00% GC)
  median time:      843.400 μs (0.00% GC)
  mean time:        856.651 μs (0.00% GC)
  maximum time:     1.047 ms (0.00% GC)
  --------------
  samples:          100
  evals/sample:     1
```

r7rohan committed:
Speedup, Test Divrem

Added speedup and tests: Divrem-Rounded

oscardssmith commented:
Have you tested speed for `BigInt`? I assume it will still be faster, but it often can cause corner cases. Other than that, this looks great!

musm commented:
> Added misc changes like: Removed useless spaces in base\div.jl

These changes are orthogonal to the PR.

r7rohan committed:
Fixed typo

+added divrem bigint test

r7rohan commented:
There was a trivial typo error fixed it.

Tested it for random BigInt, they show the same trend.
```
a, b = rand(0:500000000000000000000000000000000000000000000000000000000000000000000000000, 100000), rand(1:500000000000000000000000000000000000000000000000000000000000000000000000000, 100000);
c, d = zeros(BigInt, 100_000), zeros(BigInt, 100_000);

julia> @benchmark divrem!($a,$b,$c,$d) samples=100
BenchmarkTools.Trial:
  memory estimate:  9.92 MiB
  allocs estimate:  500000
  --------------
  minimum time:     28.940 ms (0.00% GC)
  median time:      31.278 ms (0.00% GC)
  mean time:        46.815 ms (18.10% GC)
  maximum time:     435.204 ms (51.40% GC)
  --------------
  samples:          100
  evals/sample:     1

julia> @benchmark divrem2!($a,$b,$c,$d) samples=100
BenchmarkTools.Trial:
  memory estimate:  9.92 MiB
  allocs estimate:  500000
  --------------
  minimum time:     32.511 ms (0.00% GC)
  median time:      36.163 ms (0.00% GC)
  mean time:        50.150 ms (16.95% GC)
  maximum time:     418.931 ms (53.83% GC)
  --------------
  samples:          100
  evals/sample:     1

julia> @benchmark divrem3!($a,$b,$c,$d) samples=100
BenchmarkTools.Trial:
  memory estimate:  16.02 MiB
  allocs estimate:  649972
  --------------
  minimum time:     31.629 ms (0.00% GC)
  median time:      33.585 ms (0.00% GC)
  mean time:        61.267 ms (24.99% GC)
  maximum time:     448.482 ms (51.84% GC)
  --------------
  samples:          100
  evals/sample:     1
```

Keno commented:
LGTM in the abstract, but CI is complaining about the tests.

r7rohan committed:
Fix float error

`a-d*b` is marginally off the actual value (for Float64 and BigFloat), causing some very sensitive test cases to fail.
Removed float compatibility

musm commented on test/int.jl:
```diff
@@ -339,8 +339,8 @@ let i = MyInt26779(1)
 @testset "rounding division" begin
-    for x = -100:100
-        for y = 1:100
+    for x = -50:50
```
why change these existing tests

r7rohan replied:
Probably there is no need so reverting these changes

r7rohan replied:
Probably there is no need, so reverting this change

r7rohan committed:
Revert int testset

musm commented on base/div.jl:
```diff
@@ -133,62 +132,76 @@ julia> divrem(7,3)
     if x >= 0
         if y >= 0
-            r >=        (y÷2) + (isodd(y) | iseven(q)) ? (q+true, r-y) : (q, r)
+            r >=  (y÷2) + (isodd(y) | iseven(q)) ? (q+true, r-y) : (q, r)
```
As mentioned, these are unrelated style changes. The last person who wrote this  intended to line up `(y÷2)`. Since you're not modifying the code in these methods,  it's best that  these whitespace changes  be reverted.

r7rohan committed:
Revert space

r7rohan commented:
Please merge

oscardssmith commented:
@Keno good to merge now that CI is green?

JeffBezanson commented on base/div.jl:
```diff
@@ -133,15 +133,29 @@ julia> divrem(7,3)
 """
 divrem(x, y) = divrem(x, y, RoundToZero)
+
+function divrem(a::T, b::T, r::RoundingMode) where {T<:AbstractFloat}
```
```suggestion
function divrem(a::AbstractFloat, b::AbstractFloat, r::RoundingMode)
```

Otherwise the method below will be called for different-type floats, which I think we don't want?

r7rohan replied:
I wanted to write it for Float16, Float32, Float64, BigFloat - all the subtypes of AbstractFloat. I don't know what the difference will be

r7rohan committed:
Different type-floats

Co-authored-by: Jeff Bezanson <jeff.bezanson@gmail.com>

r7rohan committed:
only int-int

- Avoids calling rem for Integers-Integers
- a-d*b not precise for Floats - AbstractFloat, AbstractIrrational.
- Rationals are still slower

Keno commented on base/div.jl:
```diff
@@ -133,15 +133,30 @@ julia> divrem(7,3)
+
+#avoids calling rem for Integers-Integers,
+#a-d*b not precise for Floats - AbstractFloat, AbstractIrrational. Rationals are still slower
+function divrem(a::Integer, b::Integer, r::RoundingMode)
```
I think this should use explicit types on `r` for the two special cases, together with the note to remove them in 2.0.

Keno commented:
This could use a PkgEval before merging. I know it doesn't seem like it, but the fallback chains for `div`-adjacent methods are very fragile because we allow packages to intercept both the 2 and 3-argument variants and want to provide various fallback. This needs to be cleaned up in 2.0, but not much we can do until then.

r7rohan committed:
Small changes

r explicitly

r7rohan committed:
small changes

r7rohan commented:
Sorry for late reply, I don't know much about PkgEval...so would appreciate help in that part. Apart from that, it seems to work well for divrem int-int types

GregPlowman commented:
Does this just need PkgEval?

oscardssmith commented:
Yeah. @Keno can you run it?

simeonschaub committed:
Merge remote-tracking branch 'origin/master' into speedup-test-divrem

simeonschaub commented:
@nanosoldier `runtests(ALL, vs = ":master")`

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/39045#issuecomment-792378088):
```
NanosoldierError: failed to run tests: Build for julia on x86_64-linux-gnu did not complete successfully

```
Logs and partial data can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_hash/24d688a_vs_5fab42a/report.md)
cc @maleadt

maleadt commented:
> curl: (6) Could not resolve host: github.com

Let's try again:

@nanosoldier `runtests(ALL, vs = ":master")`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/39045#issuecomment-792908248) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_hash/24d688a_vs_86387a8/report.md). cc @maleadt

KristofferC commented:
@nanosoldier `runtests(["AlphaStableDistributions", "BoltzmannMachines", "FilesystemDatastructures", "GPLinearODEMaker", "GeoDataFrames", "JetPackWaveFD", "Jusdl", "KissMCMC", "RiskAdjustedLinearizations", "Tar", "TensorToolbox"], vs = ":master")`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/39045#issuecomment-793504165) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_hash/24d688a_vs_faa3d41/report.md). cc @maleadt

vtjnash commented on base/div.jl:
```diff
@@ -157,6 +159,25 @@ function divrem(a, b, r::RoundingMode)
+        d = fld(a, b)
+        (d, a - d*b)
+    elseif r === RoundUp
+        # For compat. Remove in 2.0.
```
What branch is not deprecated?

vtjnash commented:
Seems to be working, but I'm confused by the line comments: why is all of the new code deprecated and marked for deletion in v2?

r7rohan commented:
@vtjnash you are free to make the changes, I'm out of touch. AFAIK its working but Keno would be able to say better.

vtjnash commented:
It seems to be working, I'm just confused by the content of the comments in the code, and was hoping you could clarify what they meant (I can make minor changes, if that's what's needed).

StefanKarpinski commented:
Bump. Clarification requested...

Keno commented:
In v2, all primitive integer types should implement three argument div(rem) and the two argument versions should just be fallbacks to those. At the moment, we have various complicated messes to handle situations where people overload the various two argument methods.

oscardssmith commented:
Is this ready to be merged? If not, what does it need?
